File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())

inputs = deque([int(input()) for _ in range(n)])
order = deque([i for i in range(1, n + 1)])


stk = []
result = []
flag = True

while order or inputs:
    while not stk or (order and stk and inputs and stk[-1] < inputs[0]):
        stk.append(order.popleft())
        result.append(1)
        
    while stk and inputs and stk[-1] == inputs[0]:
        stk.pop()
        inputs.popleft()
        result.append(0)
        
    if stk and inputs and stk[-1] > inputs[0]:
        flag = False
        break
    
        
            
if not flag:
    print("NO")
else: 
    print("\n".join(["+" if i == 1 else "-" for i in result]))
    
    

File: 백준/Silver/1874. 스택 수열/스택 수열.js
const fs = require('fs')

const [n, ...inputs] = fs.readFileSync('/dev/stdin').toString().trim().split("\n").map(c => Number(c.trim()))

const order = [...Array(n)].map((_, i) => n - i);
inputs.reverse()

const stk = []
const result = []

let flag = true;

while(!!order.length || !!inputs.length){
    while (!stk.length || (stk.length && order.length && inputs.length && stk.at(-1) < inputs.at(-1))){
        stk.push(order.pop());
        result.push(1);
    }
    
    while (stk.length && inputs.length && stk.at(-1) === inputs.at(-1)){
        stk.pop()
        inputs.pop()
        result.push(0);
    }
    
    if (stk.length && inputs.length && stk.at(-1) > inputs.at(-1)){
        flag = false;
        break;
    }
}

if (flag){
    console.log(result.map((a) => a ? "+" : "-").join("\n"))
} else {
    console.log("NO");
}
